src/base/memory_hash.c

```diff
@@ -2285,3 +2298,70 @@ mht_get_linear_hash32 (const unsigned int key, const unsigned int ht_size)
   return ret;
 }
 #endif /* ENABLE_UNUSED_FUNCTION */
+
+/*
+ * mht_put_orderly_internal
+ *   internal function for mht_put_orderly()
+ *   put data in the order.
+ *   eliminates unnecessary logic to improve performance for hash list scan.
+ *
+ *   return: key
+ *   ht(in/out): hash table (set as a side effect)
+ *   key(in): hashing key
+ *   data(in): data associated with hashing key
+ *   opt(in): options;
+ */
+static const void *
+mht_put_orderly_internal (MHT_TABLE * ht, const void *key, void *data, MHT_PUT_OPT opt)
+{
+  unsigned int hash;
+  HENTRY_PTR hentry;
+
+  assert (ht != NULL && key != NULL);
+
+  /*
+   * Hash the key and make sure that the return value is between 0 and size
+   * of hash table
+   */
+  hash = (*ht->hash_func) (key, ht->size);
+  if (hash >= ht->size)
+    {
+      hash %= ht->size;
+    }
+
+  /* This is a new entry */
+  if (ht->nprealloc_entries > 0)
+    {
+      ht->nprealloc_entries--;
+      hentry = ht->prealloc_entries;
+      ht->prealloc_entries = ht->prealloc_entries->next;
+    }
+  else
+    {
+      hentry = (HENTRY_PTR) db_fixed_alloc (ht->heap_id, DB_SIZEOF (HENTRY));
+      if (hentry == NULL)
+	{
+	  return NULL;
+	}
+    }
+
+  hentry->key = key;
+  hentry->data = data;
+
+  /* To input in order, use the act_next variable as tail. */
+  /* lru and act-related logics are not used for hash list scan, so delete them. */
+  if (ht->table[hash] == NULL)
+    {
+      ht->table[hash] = hentry;
+    }
+  else
+    {
```

Comment:
The "else" clause means that hash encounters collision but the logic including below code sequence looks cause invalid linked list as follows:
```
ht->table[hash]->act_next->next = hentry;
hentry->next = NULL;
ht->table[hash]->ac_next = hentry;
```
the above logic makes hentry be referred by both of ac_next and ac_next->next.

Reply:
I use the act_next variable as tail.
'ht->table[hash]->act_next->next = hentry;' means tail->next = hentry.
'ht->table[hash]->ac_next = hentry;' means tail is changed to hentry.
first act_next is differ from second act_next.